snap/squashfs2/squashfs2.go

```diff
@@ -0,0 +1,531 @@
+	}
+
+	sb := &squashfs_superblock{}
+	copy(sb.magic[:], data[:4])
+	if sb.magic != magic {
+		return nil, fmt.Errorf("squashfs: invalid magic")
+	}
+
+	sb.inodes = readUint32(data[4:])
+	sb.mkfsTime = readUint32(data[8:])
+	sb.blockSize = readUint32(data[12:])
+	sb.fragments = readUint32(data[16:])
+	sb.compressionType = readUint16(data[20:])
+	sb.blockSizeLog2 = readUint16(data[22:])
+	sb.flags = readUint16(data[24:])
+	sb.noIDs = readUint16(data[26:])
+	sb.s_major = readUint16(data[28:])
+	sb.s_minor = readUint16(data[30:])
+	sb.rootIno = readInodeRef(data[32:])
+	sb.bytesUsed = readInt64(data[40:])
+	sb.idTableStart = readInt64(data[48:])
+	sb.xattrIdTableSz = readInt64(data[56:])
+	sb.inodeTable = readInt64(data[64:])
+	sb.directoryTable = readInt64(data[72:])
+	sb.fragmentTable = readInt64(data[80:])
+	sb.lookupTable = readInt64(data[88:])
+	return sb, nil
+}
+
+func readSuperBlock(stream *os.File) (*squashfs_superblock, error) {
+	buffer := make([]byte, superblockSize)
+	_, err := stream.Read(buffer)
+	if err != nil {
+		return nil, err
+	}
+	return parseSuperBlock(buffer)
+}
+
+func createCompressionBackend(stream *os.File, sb *squashfs_superblock) (CompressionBackend, error) {
+	println("squashfs: compression type", sb.compressionType)
+	var optionsBlock *metaBlockReader = nil
+	if sb.flags&superBlockCompressorOptions != 0 {
+		optionsBlock = metablockReaderCreate(stream, nil, superblockSize)
+	}
+
+	switch sb.compressionType {
+	case compressionXz:
+		return createXzBackend(optionsBlock)
+	default:
+		return nil, fmt.Errorf("squashfs: unsupported compression type %d", sb.compressionType)
+	}
+}
+
+// SquashFS layout
+// ---------------
+// |  superblock   |
+// |---------------|
+// |  compression  |
+// |    options    |
+// |---------------|
+// |  datablocks   |
+// |  & fragments  |
+// |---------------|
+// |  inode table  |
+// |---------------|
+// |   directory   |
+// |     table     |
+// |---------------|
+// |   fragment    |
+// |    table      |
+// |---------------|
+// |    export     |
+// |    table      |
+// |---------------|
+// |    uid/gid    |
+// |  lookup table |
+// |---------------|
+// |     xattr     |
+// |     table     |
+// |---------------|
+func SquashFsOpen(path string) (*SquashFileSystem, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	sb, err := readSuperBlock(f)
+	if err != nil {
+		return nil, err
+	}
+
+	// handle compression type
+	cb, err := createCompressionBackend(f, sb)
+	if err != nil {
+		return nil, err
+	}
+
+	sfs := &SquashFileSystem{
+		stream:          f,
+		superBlock:      sb,
+		compression:     cb,
+		inodeReader:     metablockReaderCreate(f, cb, sb.inodeTable, sb.rootIno),
+		directoryReader: metablockReaderCreate(f, cb, sb.directoryTable),
+	}
+
+	// initialize root directory right away so we can start
+	// reading from it immediately.
+	err = sfs.loadRootDirectory()
+	if err != nil {
+		sfs.Close()
```

Comment:
Do we need error handling for "Close()" ? I.e. does it return an error, should it?

Reply:
Yes definitely, currently it just closes the stream which I'm afraid will coincide with the deferred close of the stream? Would that be an issue? But you are certainly correct here that I need to check the error